File: apps/android/app/src/main/kotlin/dev/forcetower/unes/ui/feature/calendar/components/CalCategoryGlyph.kt
package dev.forcetower.unes.ui.feature.calendar.components

import androidx.compose.foundation.Canvas
import androidx.compose.foundation.layout.size
import androidx.compose.runtime.Composable
import androidx.compose.ui.Modifier
import androidx.compose.ui.geometry.Offset
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.graphics.Path
import androidx.compose.ui.graphics.StrokeCap
import androidx.compose.ui.graphics.StrokeJoin
import androidx.compose.ui.graphics.drawscope.DrawScope
import androidx.compose.ui.graphics.drawscope.Stroke
import androidx.compose.ui.unit.Dp
import androidx.compose.ui.unit.dp
import dev.forcetower.unes.ui.feature.calendar.CalendarCategory

// Hand-rolled SVG-equivalent glyphs — three shapes: holiday (sun), exam
// (folded paper), deadline (bowtie hourglass). Mirrors `CalCatGlyph` in
// `screens-calendar.jsx`. Each shape is drawn against a 14×14 coordinate
// system and scaled to the requested size.
@Composable
internal fun CalCategoryGlyph(
    category: CalendarCategory,
    color: Color,
    size: Dp = 14.dp,
    modifier: Modifier = Modifier,
) {
    Canvas(modifier = modifier.size(size)) {
        val scale = this.size.width / 14f
        val stroke = Stroke(
            width = 1.5f * density,
            cap = StrokeCap.Round,
            join = StrokeJoin.Round,
        )
        when (category) {
            CalendarCategory.Holiday -> drawHoliday(color, scale, stroke)
            CalendarCategory.Exam -> drawExam(color, scale, stroke)
            CalendarCategory.Deadline -> drawDeadline(color, scale, stroke)
        }
    }
}

private fun DrawScope.drawHoliday(color: Color, scale: Float, stroke: Stroke) {
    drawCircle(
        color = color,
        radius = 4f * scale,
        center = Offset(7f * scale, 7f * scale),
        style = stroke,
    )
    val rays = Path()
    val pairs = listOf(
        7f to 1.5f to (7f to 3f),
        7f to 11f to (7f to 12.5f),
        1.5f to 7f to (3f to 7f),
        11f to 7f to (12.5f to 7f),
        4.2f to 4.2f to (3.1f to 3.1f),
        10.9f to 10.9f to (9.8f to 9.8f),
        4.2f to 9.8f to (3.1f to 10.9f),
        10.9f to 3.1f to (9.8f to 4.2f),
    )
    pairs.forEach { (a, b) ->
        rays.moveTo(a.first * scale, a.second * scale)
        rays.lineTo(b.first * scale, b.second * scale)
    }
    drawPath(rays, color, style = stroke)
}

private fun DrawScope.drawExam(color: Color, scale: Float, stroke: Stroke) {
    val outline = Path().apply {
        moveTo(3f * scale, 2f * scale)
        lineTo(9f * scale, 2f * scale)
        lineTo(11f * scale, 4f * scale)
        lineTo(11f * scale, 12.5f * scale)
        lineTo(2f * scale, 12.5f * scale)
        lineTo(2f * scale, 2f * scale)
        close()
    }
    drawPath(outline, color, style = stroke)
    val details = Path().apply {
        // Folded corner.
        moveTo(9f * scale, 2f * scale)
        lineTo(9f * scale, 4.5f * scale)
        lineTo(11f * scale, 4.5f * scale)
        // Three rule lines.
        listOf(7f to 5f, 9.5f to 5f, 12f to 3f).forEach { (y, len) ->
            moveTo(4.5f * scale, y * scale)
            lineTo((4.5f + len) * scale, y * scale)
        }
    }
    drawPath(details, color, style = stroke)
}

private fun DrawScope.drawDeadline(color: Color, scale: Float, stroke: Stroke) {
    val bars = Path().apply {
        moveTo(3.5f * scale, 2f * scale)
        lineTo(10.5f * scale, 2f * scale)
        moveTo(3.5f * scale, 12f * scale)
        lineTo(10.5f * scale, 12f * scale)
    }
    drawPath(bars, color, style = stroke)
    val curves = Path().apply {
        moveTo(4.5f * scale, 2f * scale)
        quadraticTo(4.5f * scale, 7f * scale, 9.5f * scale, 12f * scale)
        moveTo(9.5f * scale, 2f * scale)
        quadraticTo(9.5f * scale, 7f * scale, 4.5f * scale, 12f * scale)
    }
    drawPath(curves, color, style = stroke)
}

File: apps/ios/UNES/Features/Calendar/Components/CalCategoryGlyph.swift
import SwiftUI

/// Hand-rolled SVG-equivalent glyph used in the eyebrow row of every calendar
/// surface. Mirrors `CalCatGlyph` in `screens-calendar.jsx`.
///
/// Three shapes:
///   - **holiday** — sun (circle + 8 rays)
///   - **exam**    — folded paper with rule-lines
///   - **deadline** — bowtie / hourglass crossings
struct CalCategoryGlyph: View {
    let category: CalendarCategory
    var color: Color
    var size: CGFloat = 14

    var body: some View {
        switch category {
        case .holiday:  HolidayGlyph(color: color, size: size)
        case .exam:     ExamGlyph(color: color, size: size)
        case .deadline: DeadlineGlyph(color: color, size: size)
        }
    }
}

private struct HolidayGlyph: View {
    let color: Color
    let size: CGFloat

    var body: some View {
        let stroke = StrokeStyle(lineWidth: 1.5, lineCap: .round, lineJoin: .round)
        Canvas { ctx, _ in
            let s = size / 14
            // Sun centre at (7,7), r=4. Rays at cardinals + diagonals.
            let center = CGPoint(x: 7 * s, y: 7 * s)
            let circle = Path(ellipseIn: CGRect(x: 3 * s, y: 3 * s, width: 8 * s, height: 8 * s))
            ctx.stroke(circle, with: .color(color), style: stroke)

            // Cardinals: top, bottom, left, right (length 1.5 inset by 0.5)
            let pairs: [(CGPoint, CGPoint)] = [
                (CGPoint(x: 7,   y: 1.5), CGPoint(x: 7,   y: 3)),
                (CGPoint(x: 7,   y: 11),  CGPoint(x: 7,   y: 12.5)),
                (CGPoint(x: 1.5, y: 7),   CGPoint(x: 3,   y: 7)),
                (CGPoint(x: 11,  y: 7),   CGPoint(x: 12.5, y: 7)),
                // Diagonals: 4.2 4.2 → 3.1 3.1 etc.
                (CGPoint(x: 4.2, y: 4.2), CGPoint(x: 3.1, y: 3.1)),
                (CGPoint(x: 10.9, y: 10.9), CGPoint(x: 9.8, y: 9.8)),
                (CGPoint(x: 4.2, y: 9.8), CGPoint(x: 3.1, y: 10.9)),
                (CGPoint(x: 10.9, y: 3.1), CGPoint(x: 9.8, y: 4.2)),
            ]
            var rays = Path()
            for (a, b) in pairs {
                rays.move(to: CGPoint(x: a.x * s, y: a.y * s))
                rays.addLine(to: CGPoint(x: b.x * s, y: b.y * s))
            }
            ctx.stroke(rays, with: .color(color), style: stroke)
            _ = center
        }
        .frame(width: size, height: size)
    }
}

private struct ExamGlyph: View {
    let color: Color
    let size: CGFloat

    var body: some View {
        let stroke = StrokeStyle(lineWidth: 1.5, lineCap: .round, lineJoin: .round)
        Canvas { ctx, _ in
            let s = size / 14
            // Folded paper outline: 3,2 → 9,2 → 11,4 → 11,12.5 → 2,12.5 → 2,2 → close
            var outline = Path()
            outline.move(to: CGPoint(x: 3 * s, y: 2 * s))
            outline.addLine(to: CGPoint(x: 9 * s, y: 2 * s))
            outline.addLine(to: CGPoint(x: 11 * s, y: 4 * s))
            outline.addLine(to: CGPoint(x: 11 * s, y: 12.5 * s))
            outline.addLine(to: CGPoint(x: 2 * s, y: 12.5 * s))
            outline.addLine(to: CGPoint(x: 2 * s, y: 2 * s))
            outline.closeSubpath()
            ctx.stroke(outline, with: .color(color), style: stroke)

            // Fold + rule lines.
            var details = Path()
            // Fold: 9,2 → 9,4.5 → 11,4.5
            details.move(to: CGPoint(x: 9 * s, y: 2 * s))
            details.addLine(to: CGPoint(x: 9 * s, y: 4.5 * s))
            details.addLine(to: CGPoint(x: 11 * s, y: 4.5 * s))
            // Three rule lines.
            for (y, len): (CGFloat, CGFloat) in [(7, 5), (9.5, 5), (12, 3)] {
                details.move(to: CGPoint(x: 4.5 * s, y: y * s))
                details.addLine(to: CGPoint(x: (4.5 + len) * s, y: y * s))
            }
            ctx.stroke(details, with: .color(color), style: stroke)
        }
        .frame(width: size, height: size)
    }
}

private struct DeadlineGlyph: View {
    let color: Color
    let size: CGFloat

    var body: some View {
        let stroke = StrokeStyle(lineWidth: 1.5, lineCap: .round, lineJoin: .round)
        Canvas { ctx, _ in
            let s = size / 14
            // Top + bottom bars
            var bars = Path()
            bars.move(to: CGPoint(x: 3.5 * s, y: 2 * s))
            bars.addLine(to: CGPoint(x: 10.5 * s, y: 2 * s))
            bars.move(to: CGPoint(x: 3.5 * s, y: 12 * s))
            bars.addLine(to: CGPoint(x: 10.5 * s, y: 12 * s))
            ctx.stroke(bars, with: .color(color), style: stroke)

            // Crossing curves: simulate the hourglass-ish quadratic curves.
            var curves = Path()
            curves.move(to: CGPoint(x: 4.5 * s, y: 2 * s))
            curves.addQuadCurve(to: CGPoint(x: 9.5 * s, y: 12 * s),
                                control: CGPoint(x: 4.5 * s, y: 7 * s))
            curves.move(to: CGPoint(x: 9.5 * s, y: 2 * s))
            curves.addQuadCurve(to: CGPoint(x: 4.5 * s, y: 12 * s),
                                control: CGPoint(x: 9.5 * s, y: 7 * s))
            ctx.stroke(curves, with: .color(color), style: stroke)
        }
        .frame(width: size, height: size)
    }
}
